File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/grpc/ClientInterceptorAdapter.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.arrow.flight.grpc;

import java.util.ArrayList;
import java.util.List;

import org.apache.arrow.flight.CallInfo;
import org.apache.arrow.flight.CallStatus;
import org.apache.arrow.flight.FlightClientMiddleware;
import org.apache.arrow.flight.FlightMethod;
import org.apache.arrow.flight.FlightRuntimeException;
import org.apache.arrow.flight.FlightStatusCode;

import io.grpc.CallOptions;
import io.grpc.Channel;
import io.grpc.ClientCall;
import io.grpc.ClientInterceptor;
import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
import io.grpc.ForwardingClientCallListener.SimpleForwardingClientCallListener;
import io.grpc.Metadata;
import io.grpc.MethodDescriptor;
import io.grpc.Status;
import io.grpc.StatusRuntimeException;

/**
 * An adapter between Flight client middleware and gRPC interceptors.
 *
 * <p>This is implemented as a single gRPC interceptor that runs all Flight client middleware sequentially.
 */
public class ClientInterceptorAdapter implements ClientInterceptor {

  private final List<FlightClientMiddleware.Factory> factories;

  public ClientInterceptorAdapter(List<FlightClientMiddleware.Factory> factories) {
    this.factories = factories;
  }

  @Override
  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
      CallOptions callOptions, Channel next) {
    final List<FlightClientMiddleware> middleware = new ArrayList<>();
    final CallInfo info = new CallInfo(FlightMethod.fromProtocol(method.getFullMethodName()));

    try {
      for (final FlightClientMiddleware.Factory factory : factories) {
        middleware.add(factory.onCallStarted(info));
      }
    } catch (FlightRuntimeException e) {
      // Explicitly propagate
      throw e;
    } catch (StatusRuntimeException e) {
      throw StatusUtils.fromGrpcRuntimeException(e);
    } catch (RuntimeException e) {
      throw StatusUtils.fromThrowable(e);
    }
    return new FlightClientCall<>(next.newCall(method, callOptions), middleware);
  }

  /**
   * The ClientCallListener which hooks into the gRPC request cycle and actually runs middleware at certain points.
   */
  private static class FlightClientCallListener<RespT> extends SimpleForwardingClientCallListener<RespT> {

    private final List<FlightClientMiddleware> middleware;
    boolean receivedHeaders;

    public FlightClientCallListener(ClientCall.Listener<RespT> responseListener,
        List<FlightClientMiddleware> middleware) {
      super(responseListener);
      this.middleware = middleware;
      receivedHeaders = false;
    }

    @Override
    public void onHeaders(Metadata headers) {
      receivedHeaders = true;
      final MetadataAdapter adapter = new MetadataAdapter(headers);
      try {
        middleware.forEach(m -> m.onHeadersReceived(adapter));
      } finally {
        // Make sure to always call the gRPC callback to avoid interrupting the gRPC request cycle
        super.onHeaders(headers);
      }
    }

    @Override
    public void onClose(Status status, Metadata trailers) {
      try {
        if (!receivedHeaders) {
          // gRPC doesn't always send response headers if the call errors or completes immediately, but instead
          // consolidates them with the trailers. If we never got headers, assume this happened and run the header
          // callback with the trailers.
          final MetadataAdapter adapter = new MetadataAdapter(trailers);
          middleware.forEach(m -> m.onHeadersReceived(adapter));
        }
        final CallStatus flightStatus = StatusUtils.fromGrpcStatusAndTrailers(status, trailers);
        middleware.forEach(m -> m.onCallCompleted(flightStatus));
      } finally {
        // Make sure to always call the gRPC callback to avoid interrupting the gRPC request cycle
        super.onClose(status, trailers);
      }
    }
  }

  /**
   * The gRPC ClientCall which hooks into the gRPC request cycle and injects our ClientCallListener.
   */
  private static class FlightClientCall<ReqT, RespT> extends SimpleForwardingClientCall<ReqT, RespT> {

    private final List<FlightClientMiddleware> middleware;

    public FlightClientCall(ClientCall<ReqT, RespT> clientCall, List<FlightClientMiddleware> middleware) {
      super(clientCall);
      this.middleware = middleware;
    }

    @Override
    public void start(Listener<RespT> responseListener, Metadata headers) {
      final MetadataAdapter metadataAdapter = new MetadataAdapter(headers);
      middleware.forEach(m -> m.onBeforeSendingHeaders(metadataAdapter));

      super.start(new FlightClientCallListener<>(responseListener, middleware), headers);
    }

    @Override
    public void cancel(String message, Throwable cause) {
      final CallStatus flightStatus = new CallStatus(FlightStatusCode.CANCELLED, cause, message, null);
      middleware.forEach(m -> m.onCallCompleted(flightStatus));
      super.cancel(message, cause);
    }
  }
}

File: csharp/src/Apache.Arrow.Flight/Middleware/Interceptors/ClientInterceptorAdapter.cs
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements. See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Apache.Arrow.Flight.Middleware.Interfaces;
using Grpc.Core;
using Grpc.Core.Interceptors;

namespace Apache.Arrow.Flight.Middleware.Interceptors;

public sealed class ClientInterceptorAdapter : Interceptor
{
    private readonly IReadOnlyList<IFlightClientMiddlewareFactory> _factories;

    public ClientInterceptorAdapter(IEnumerable<IFlightClientMiddlewareFactory> factories)
    {
        _factories = factories?.ToList() ?? throw new ArgumentNullException(nameof(factories));
    }

    public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(
        TRequest request,
        ClientInterceptorContext<TRequest, TResponse> context,
        AsyncUnaryCallContinuation<TRequest, TResponse> continuation)
        where TRequest : class
        where TResponse : class
    {
        var options = InterceptCall(context, out var middlewares);

        var newContext = new ClientInterceptorContext<TRequest, TResponse>(
            context.Method,
            context.Host,
            options);

        var call = continuation(request, newContext);

        return new AsyncUnaryCall<TResponse>(
            HandleResponse(call.ResponseAsync, call.ResponseHeadersAsync, call.GetStatus, call.GetTrailers,
                call.Dispose, middlewares),
            call.ResponseHeadersAsync,
            call.GetStatus,
            call.GetTrailers,
            call.Dispose
        );
    }

    public override AsyncServerStreamingCall<TResponse> AsyncServerStreamingCall<TRequest, TResponse>(
        TRequest request,
        ClientInterceptorContext<TRequest, TResponse> context,
        AsyncServerStreamingCallContinuation<TRequest, TResponse> continuation)
        where TRequest : class
        where TResponse : class
    {
        var callOptions = InterceptCall(context, out var middlewares);
        var newContext = new ClientInterceptorContext<TRequest, TResponse>(
            context.Method, context.Host, callOptions);

        var call = continuation(request, newContext);

        var responseHeadersTask = call.ResponseHeadersAsync.ContinueWith(task =>
        {
            if (task.IsFaulted)
            {
                throw task.Exception!;
            }
            
            if (task.IsCanceled)
            {
                throw new TaskCanceledException(task);
            }
            
            var headers = task.Result;
            var ch = new CallHeaders(headers);
            foreach (var m in middlewares)
                m?.OnHeadersReceived(ch);

            return headers;
        });

        var wrappedResponseStream = new MiddlewareResponseStream<TResponse>(
            call.ResponseStream,
            call,
            middlewares);

        return new AsyncServerStreamingCall<TResponse>(
            wrappedResponseStream,
            responseHeadersTask,
            call.GetStatus,
            call.GetTrailers,
            call.Dispose);
    }


    private CallOptions InterceptCall<TRequest, TResponse>(
        ClientInterceptorContext<TRequest, TResponse> context,
        out List<IFlightClientMiddleware> middlewareList)
        where TRequest : class
        where TResponse : class
    {
        var callInfo = new CallInfo(context.Method.FullName, context.Method.Type);

        var headers = context.Options.Headers ?? new Metadata();
        middlewareList = new List<IFlightClientMiddleware>();

        var callHeaders = new CallHeaders(headers);

        foreach (var factory in _factories)
        {
            var middleware = factory.OnCallStarted(callInfo);
            middleware?.OnBeforeSendingHeaders(callHeaders);
            middlewareList.Add(middleware);
        }

        return context.Options.WithHeaders(headers);
    }

    private async Task<TResponse> HandleResponse<TResponse>(
        Task<TResponse> responseTask,
        Task<Metadata> headersTask,
        Func<Status> getStatus,
        Func<Metadata> getTrailers,
        Action dispose,
        List<IFlightClientMiddleware> middlewares)
    {
        var nonNullMiddlewares = (middlewares ?? new List<IFlightClientMiddleware>())
            .Where(m => m != null)
            .ToList();

        var hasMiddlewares = nonNullMiddlewares.Count > 0;
        var completionNotified = false;
        
        try
        {
            // Always await headers to surface faults; only materialize CallHeaders if needed.
            var headers = await headersTask.ConfigureAwait(false);
            if (hasMiddlewares)
            {
                var ch = new CallHeaders(headers);
                foreach (var m in nonNullMiddlewares)
                    m.OnHeadersReceived(ch);
            }

            var response = await responseTask.ConfigureAwait(false);

            // Single completion notification
            NotifyCompletionOnce();
            return response;
        }
        catch
        {
            // Completion on failure (only once)
            NotifyCompletionOnce();
            throw;
        }
        finally
        {
            dispose?.Invoke();
        }
        
        void NotifyCompletionOnce()
        {
            if (completionNotified || !hasMiddlewares) return;
            completionNotified = true;

            var status = getStatus();
            var trailers = getTrailers();

            foreach (var m in nonNullMiddlewares)
                m.OnCallCompleted(status, trailers);
        }
    }
}